### PR TITLE
Tcp server

### PR DESCRIPTION
**Pull request by t-mon**

### CODE REVIEW

File: libnymea-remoteproxy/authentication/authenticationreply.cpp [hunk 1]
```diff
@@ -20,23 +20,30 @@
...
     connect(m_timer, &QTimer::timeout, this, &AuthenticationReply::onTimeout);
-
+    qCDebug(dcAuthentication) << "Created authentication reply for" << proxyClient << "Timeout:" << Engine::instance()->configuration()->authenticationTimeout() << "[ms]";
     m_timer->start(Engine::instance()->configuration()->authenticationTimeout());
 }
 
-ProxyClient *AuthenticationReply::proxyClient() const
+AuthenticationReply::~AuthenticationReply()
+{
+    qCCritical(dcAuthentication()) << "Destroy authentication reply";
```
Comment (mzanetti) [inline libnymea-remoteproxy/authentication/authenticationreply.cpp]: drop that

File: libnymea-remoteproxy/jsonrpcserver.cpp [hunk 2]
```diff
@@ -316,6 +246,99 @@ void JsonRpcServer::processData(ProxyClient *proxyClient, const QByteArray &data
...
+    qCDebug(dcJsonRpcTraffic()) << "Incoming data from" << proxyClient << ": " << data;
+
+    // Handle package fragmentation
+    QList<QByteArray> packages = proxyClient->processData(data);
+
+    // Make sure the buffer size is in range
+    if (proxyClient->bufferSizeViolation()) {
+        qCWarning(dcJsonRpc()) << "Data buffer size violation from" << proxyClient;
+        proxyClient->killConnection("Data buffer size violation. Data >= 10 kB");
+        return;
```
Comment (mzanetti) [inline libnymea-remoteproxy/jsonrpcserver.cpp]: IMHO this isn't nice... bufferSizeViolation() is checked for >= 10KB inside proxyClient, yet the actual decision whether to disconnect it or not, and even worse the debug print for what the size limit is, is made in the server... It should either be:

if (proxyClient->bufferSize() >= 10KB) {
  qCWarning() << ...;
  proxyClient->kill(...);
}

or just

proxyClient->validateOrKill();

and all the magic for it is inside proxyClient.... But don't distribute a single logic across different classes.

**Comment by mzanetti** [comment]
Observing occational test failures.

`FAIL!  : RemoteProxyOfflineTests::tcpRemoteConnection() 'remoteConnectionDataTwo.count() == 1' returned FALSE. ()
   Loc: [nymea-remoteproxy-tests-offline.cpp(1255)]`

**Comment by mzanetti** [comment]
Crashes on incoming connections:

`Jul 01 13:02:02 nymea nymead[1244]:  I | AWS: Proxy remote connection request received
Jul 01 13:02:02 nymea nymead[1244]:  I | Cloud: Connecting to remote proxy server "wss://remoteproxy.nymea.io"
Jul 01 13:02:02 nymea nymead[1244]:  I | RemoteProxyClientConnection: Creating a web socket connection to "wss://remoteproxy.nymea.io"
Jul 01 13:02:02 nymea nymead[1244]:  I | RemoteProxyClientConnection: Connecting to "wss://remoteproxy.nymea.io"
Jul 01 13:02:02 nymea nymead[1244]:  I | RemoteProxyClientWebSocket: Connecting to "wss://remoteproxy.nymea.io"
Jul 01 13:02:02 nymea nymead[1244]:  I | RemoteProxyClientWebSocket: Socket state changed QAbstractSocket::ConnectingState
Jul 01 13:02:02 nymea nymead[1244]:  W | Application: Cought SIGSEGV signal. Segmentation fault!
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "[1] /lib/arm-linux-gnueabihf/libc.so.6"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    ()+0x25750"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    ??:0"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "[2] /usr/lib/arm-linux-gnueabihf/libQt5Core.so.5"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    QMetaObject::indexOfEnumerator(char const*) const+0x2d"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    ??:0"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "[3] /usr/lib/arm-linux-gnueabihf/libQt5Core.so.5"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    qt_QMetaEnum_debugOperator(QDebug&, int, QMetaObject const*, char const*)+0x2f"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    ??:0"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "[4] /usr/lib/arm-linux-gnueabihf/libnymea-remoteproxyclient.so.1"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    remoteproxyclient::RemoteProxyConnection::setState(remoteproxyclient::RemoteProxyConnection::State)+0xb7"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    ??:0"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "[5] /usr/lib/arm-linux-gnueabihf/libQt5Core.so.5"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    QMetaObject::activate(QObject*, int, int, void**)+0x4b9"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    ??:0"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "[6] /usr/lib/arm-linux-gnueabihf/libnymea-remoteproxyclient.so.1"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    remoteproxyclient::ProxyConnection::stateChanged(QAbstractSocket::SocketState)+0x25"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    ??:0"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "[7] /usr/lib/arm-linux-gnueabihf/libnymea-remoteproxyclient.so.1"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    remoteproxyclient::WebSocketConnection::onStateChanged(QAbstractSocket::SocketState)+0x35"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    ??:0"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "[8] /usr/lib/arm-linux-gnueabihf/libnymea-remoteproxyclient.so.1"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    ()+0x178d6"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    ??:0"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "[9] /usr/lib/arm-linux-gnueabihf/libQt5Core.so.5"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    QMetaObject::activate(QObject*, int, int, void**)+0x40f"
Jul 01 13:02:02 nymea nymead[1244]:  C | Application: "    ??:0"
Jul 01 13:02:03 nymea systemd[1]: nymead.service: Main process exited, code=dumped, status=11/SEGV
Jul 01 13:02:03 nymea systemd[1]: nymead.service: Failed with result 'core-dump'.
`

**Review by mzanetti — CHANGES_REQUESTED** [review]

File: debian/rules [hunk 3]
```diff
@@ -15,6 +15,9 @@ override_dh_strip:
 	dh_strip --dbg-package=libnymea-remoteproxy-dbg
 	dh_strip --dbg-package=libnymea-remoteproxyclient-dbg
 
+#override_dh_auto_test:
+#	make test
+
```
[the remark this answers is not in the document]
Comment (t-mon) [reply debian/rules:20]: dropped in #23

File: libnymea-remoteproxy/proxyclient.cpp [hunk 4]
```diff
@@ -190,6 +211,37 @@ void ProxyClient::killConnection(const QString &reason)
...
+    int splitIndex = m_dataBuffers.indexOf("}\n{");
+    while (splitIndex > -1) {
+        packages.append(m_dataBuffers.left(splitIndex + 1));
+        m_dataBuffers = m_dataBuffers.right(m_dataBuffers.length() - splitIndex - 2);
+        splitIndex = m_dataBuffers.indexOf("}\n{");
+    }
+    if (m_dataBuffers.trimmed().endsWith("}")) {
+        packages.append(m_dataBuffers);
+        m_dataBuffers.clear();
+    }
```
Comment (mzanetti) [inline libnymea-remoteproxy/proxyclient.cpp:235]: Considering this input buffer:

`{ "id": 1, ... }\n{ "id": 2, "someMap": { "a": "b" }` split `}\n`

Wouldn't it return the packet 2 in the buffer even though it's not complete yet?

I guess 232 shouldn't trim the buffer and check for `endsWith("}\n")` instead?

File: libnymea-remoteproxyclient/proxyjsonrpcclient.cpp [hunk 5]
```diff
@@ -119,4 +117,22 @@ void JsonRpcClient::processData(const QByteArray &data)
...
+    int splitIndex = m_dataBuffer.indexOf("}\n{");
+    while (splitIndex > -1) {
+        processDataPackage(m_dataBuffer.left(splitIndex + 1));
+        m_dataBuffer = m_dataBuffer.right(m_dataBuffer.length() - splitIndex - 2);
+        splitIndex = m_dataBuffer.indexOf("}\n{");
+    }
+    if (m_dataBuffer.trimmed().endsWith("}")) {
+        processDataPackage(m_dataBuffer);
+        m_dataBuffer.clear();
+    }
```
Comment (mzanetti) [inline libnymea-remoteproxyclient/proxyjsonrpcclient.cpp:135]: Why is this code here a second time?
ProxyClient::processData() and JsonRpcClient::processData(), doing the same thing

File: libnymea-remoteproxy/proxyclient.cpp [hunk 6]
```diff
@@ -190,6 +211,37 @@ void ProxyClient::killConnection(const QString &reason)
...
+
+    return packages;
```
Comment (mzanetti) [inline libnymea-remoteproxy/proxyclient.cpp:237]: Package / packet

File: libnymea-remoteproxy/proxyclient.h [hunk 7]
```diff
@@ -105,6 +119,11 @@ class ProxyClient : public QObject
 
     QString m_userName;
 
+    // Json data information
+    int m_messageId = 0;
+    QByteArray m_dataBuffers;
+    bool m_bufferSizeViolation = false;
```
Comment (mzanetti) [inline libnymea-remoteproxy/proxyclient.h:125]: defined but not used

File: libnymea-remoteproxy/tcpsocketserver.h [hunk 8]
```diff
@@ -0,0 +1,100 @@
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
+*
+*  Copyright 2013 - 2020, nymea GmbH
```
Comment (mzanetti) [inline libnymea-remoteproxy/tcpsocketserver.h:3]: 22

File: libnymea-remoteproxy/tcpsocketserver.h [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    void sendData(const QUuid &clientId, const QByteArray &data) override;
+    void killClientConnection(const QUuid &clientId, const QString &killReason) override;
+
+    bool running() const override;
+
+private:
+    bool m_sslEnabled;
+    QSslConfiguration m_sslConfiguration;
+
+     QHash<QUuid, QTcpSocket *> m_clientList;
```
Comment (mzanetti) [inline libnymea-remoteproxy/tcpsocketserver.h:83]: ` `

File: libnymea-remoteproxy/websocketserver.cpp [hunk 10]
```diff
@@ -93,12 +84,16 @@ void WebSocketServer::onClientConnected()
 {
     // Got a new client connected
     QWebSocket *client = m_server->nextPendingConnection();
+    if (!client) {
+        qCWarning(dcWebSocketServer()) << "Next pending connection dissapeared. Doing nothing.";
```
Comment (mzanetti) [inline libnymea-remoteproxy/websocketserver.cpp:88]: disappeared (or vanished)

File: libnymea-remoteproxy/websocketserver.cpp [hunk 11]
```diff
@@ -149,7 +144,10 @@ void WebSocketServer::onBinaryMessageReceived(const QByteArray &data)
 void WebSocketServer::onClientError(QAbstractSocket::SocketError error)
 {
     QWebSocket *client = static_cast<QWebSocket *>(sender());
-    qCWarning(dcWebSocketServer()) << "Client error occurred:" << error << client->errorString();
+    qCWarning(dcWebSocketServer()) << "Client error occurred:" << client << client->peerAddress().toString() << error << client->errorString() << "Closing the socket.";
+
+    // Note: on any error which can occure, make sure the socket will be closed in any case
```
Comment (mzanetti) [inline libnymea-remoteproxy/websocketserver.cpp:149]: occur

File: libnymea-remoteproxy/websocketserver.cpp [hunk 12]
```diff
@@ -188,16 +190,18 @@ bool WebSocketServer::stopServer()
...
-    if (m_server) {
-        qCDebug(dcWebSocketServer()) << "Stop server" << m_server->serverName() << serverUrl().toString();
-        m_server->close();
-        delete m_server;
-        m_server = nullptr;
-    }
+    if (!m_server)
+        return true;
 
+    qCDebug(dcWebSocketServer()) << "Stop server" << m_server->serverName() << serverUrl().toString();
```
Comment (mzanetti) [inline libnymea-remoteproxy/websocketserver.cpp:201]: Stopping

File: libnymea-remoteproxyclient/proxyjsonrpcclient.cpp [hunk 13]
```diff
@@ -73,10 +73,8 @@ void JsonRpcClient::sendRequest(const QVariantMap &request)
     m_connection->sendData(data);
 }
 
-void JsonRpcClient::processData(const QByteArray &data)
+void JsonRpcClient::processDataPackage(const QByteArray &data)
```
Comment (mzanetti) [inline libnymea-remoteproxyclient/proxyjsonrpcclient.cpp:76]: packet

File: nymea-remoteproxy.pri [hunk 14]
```diff
@@ -5,7 +5,7 @@ QT -= gui
 SERVER_NAME=nymea-remoteproxy
 API_VERSION_MAJOR=0
 API_VERSION_MINOR=3
-SERVER_VERSION=0.1.7
+SERVER_VERSION=0.1.8
 
 DEFINES += SERVER_NAME_STRING=\\\"$${SERVER_NAME}\\\" \
            SERVER_VERSION_STRING=\\\"$${SERVER_VERSION}\\\" \
```
Comment (mzanetti) [inline nymea-remoteproxy.pri:11]: Would be nicer to use a config.in file for this, as passing DEFINES will invalidate the entire compiler cache each time you switch a branch with a different version.

Would be just creating a version.in.h file with  `#define SERVER_VERSION_STRING '"$$SERVER_VERSION_STRING"'`

and use `QMAKE_SUBSTITUTES += version.h.in` here
 Then include version.h where the string is needed and ccache will only recompile that one file

File: server/main.cpp [hunk 15]
```diff
@@ -145,7 +116,7 @@ int main(int argc, char *argv[])
                                              "registered nymea deamons to establish a tunnel connection.\n\n"
                                              "Version: %1\n"
                                              "API version: %2\n\n"
-                                             "Copyright %3 2018 Simon Stürz <simon.stuerz@guh.io>\n")
+                                             "Copyright %3 2021 nymea GmbH <developer@nymea.io>\n")
```
Comment (mzanetti) [inline server/main.cpp:119]: 22

File: tests/test-offline/nymea-remoteproxy-tests-offline.cpp [hunk 16]
```diff
@@ -286,13 +334,70 @@ void RemoteProxyOfflineTests::websocketPing()
     stopServer();
 }
 
+//void RemoteProxyOfflineTests::apiBasicCallsTcp_data()
```
Comment (mzanetti) [inline tests/test-offline/nymea-remoteproxy-tests-offline.cpp:337]: not finished yet?

**Review by mzanetti — CHANGES_REQUESTED** [review]

**Pull request merged**